### ISSUE TITLE
Escaping issue with LP_ENABLE_BATT and git repo on OSX+iterm+bash

### ISSUE THREAD

**Issue by HartS** [nojhan/liquidprompt#661]
I'm not sure if anything here is tied to my environment, and it's worth noting that homebrew is still installing Liquidprompt 1.12, so it's possible this has actually been fixed in recent liquidprompt versions. But if it has been fixed, I couldn't find mention of this in closed issues, and the homebrew package is unfortunately lagging

When using liquidprompt with the default prompt values, I've been experiencing a really frustrating issue with reverse-i search inside of a git repository. The issue doesn't seem to be related to *just* the `LP_ENABLE_BATT` setting, as it doesn't occur outside of git repos, but setting `LP_ENABLE_BATT=0` *does* resolve it (which is my current workaround, as battery display isn't super important to me, and git repo support is)

OSX: 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64
Bash: 3.2.57(1)-release
LP version: 1.12

### Current Behavior
(with default values set in liquidpromptrc)
Step 1: cd into a git repo e.g. `cd Desktop/development/git/liquidprompt/`
Prompt will display: `⏚ [hartsimha:~/Desktop/development/git/liquidprompt] master ±`
Step 2: `ctrl+r` # for reverse-i search
Step 3: type `cd` # find command from step 1 in reverse-i search
Step 4: `ctrl+a` # end reverse-i search, select command from step 1, move to beginning of line
Prompt will now display `⏚ [hartsimha:~/Desktop/development/git/liquidprompt] master ± cd Desktop/development/git/liquidprompt/`, and the cursor will be displayed on the `±` character, instead of the `c` character

Alternately, if you use `ctrl+e` in step 4, to go to the end of the line, the cursor will be displayed on the `t`, 2 characters behind its actual position.

This is really problematic when using reverse-i search to bring up previous commands and edit them

### Possible Solution
Right now I've disabled the battery display with `LP_ENABLE_BATT=1` in my `.liquidpromptrc` and that's resolved the issue for me. However, this is just a workaround for the underlying escaping issue.

**Comment by Rycieos** [comment]
> the homebrew package is unfortunately lagging

Yes, we are working on updating it, but are running into issues with the Homebrew project.

> homebrew is still installing Liquidprompt 1.12

The default version is, but you can select a version that tracks with the `master` branch. I don't use Homebrew, so I don't know how that works exactly, but the [formula page](https://formulae.brew.sh/formula/liquidprompt) lists a `head` version.

This looks like possibly a duplicate to #351 and #500? And maybe similar to #511 and #547.

If it is related, then this is an unresolved issue for a while that I cannot track down. It also hasn't helped that the original reporters have stopped responding, and my attempts to reproduce it leave me with more questions than answers. #547 has most of my notes on the issue.

I will 100% work with you to fix this, but I can't make any promises that it is actually fixable. I'm thinking right now it is a Bash bug, but more digging is needed.

All that said, something to try: if you set `LP_MARK_ADAPTER` to an ASCII character (eg `A`), does the issue go away?
Same thing with `LP_MARK_GIT`, set to a non-Unicode character. Try multiple combinations, see what happens.

**Comment by HartS** [comment]
Thank you! I'll give these a try tonight, and let me know if there's anything I can look up also.

I took a look at the other issues, and all but #500 look related to long lines wrapping. In my case, lines were not being wrapped at any point. I just went through the repro steps again to verify

**Comment by augmentedfourth** [comment]
To switch to the HEAD/master release, use this command:
`brew uninstall liquidprompt && brew install --HEAD liquidprompt`

Also, to upgrade in the future, you need to use `brew upgrade --fetch-HEAD` when upgrading liquidprompt (or just use that as your regular `brew upgrade` command, perhaps with an alias or shell function).

I have a PR open to get the liquidprompt 2 release into Homebrew, but I haven't had enough time to devote to building the kind of installation test they'll deem appropriate. Part of the difficulty is that the existing installation test seems to have been created by automation and doesn't perform an accurate test, and replicating that action with the current version is being deemed unsatisfactory by the Homebrew maintainers.

**Comment by Rycieos** [comment]
I cannot reproduce this with versions v1.12.0, v1.12.1, or d5e8f5afdc7555bc3606b742aeaac7dc7c117f3a on my Linux machine with Bash 3.2 and Bash 4.4. But I can reproduce this with d5e8f5afdc7555bc3606b742aeaac7dc7c117f3a on a MacOS X 10.15 (kernel 19.3.0) VM.

I can also confirm it happens on both Terminal and iTerm 2. And only in Bash, not Zsh. So it seems specific to MacOS's default version of Bash 3.2.

I'm assuming it is a problem with multi-byte characters, but I don't see why. MacOS X is 100% UTF-8, and it seems that their Bash 3.2 is as well.

But this curiosity showed up:
```bash
$ printf '%q\n' '⏚'
$'�\217\232'

$ printf '%s\n' $'\342\217\232'
⏚
```

So I can print the character from its bytes, but when I ask Bash to do the same, it mangles or loses the first byte. This is the only thing "wrong" I can find so far. And this seems to be the case for any 3 byte character. Bash 3.2 on my Linux machine prints the byte sequence correctly. And later versions of Bash just print the character itself.

Weirdly, if I set `LP_MARK_ADAPTER=±`, the cursor goes one char to the right; that is, between the "±" and "cd". So it is not limited to 3 byte characters, only a 2-byte character will trigger it as well. And I only need one, see below. So the above finding is probably nothing.

This is as reduced as I can make it:
```bash
PS1='± userdev/liquidprompt $ '  # breaks
PS1=' ± userdev/liquidpromp $ '  # breaks
PS1='± userdev/liquidpromp $ '   # does not (only 1 character shorter!)
```
From what I can tell, if a multi-byte character is one of the first two chars printed in PS1, and printed PS1 is at least 25 characters long, it breaks.

This seems solidly into the realm of buggy Bash to me, but I'm open to ideas.

**Comment by HartS** [comment]
Wow, thanks for the in-depth investigation! To follow up with the suggestion from yesterday, I did try setting `LP_MARK_ADAPTER=A`, `LP_MARK_GIT=A`, and combinations. `LP_MARK_GIT` didn't seem to make a difference (I could still see the issue inside of a git repo, but not outside of one).

`LP_MARK_ADAPTER=A` *also* resolves the issue, however.

Would you like me to try upgrading bash? I have a mind to do so now anyway, and if upgrading bash resolves the issue, I'm inclined to chalk this up to a bug in bash 3.2 rather than liquidprompt (some doc note about incompatibility may assist other OSX users though)

**Comment by Rycieos** [comment]
> Would you like me to try upgrading bash?

Sure, we could isolate the issue to MacOS Bash 3.2, instead of MacOS Bash.

> some doc note about incompatibility may assist other OSX users though

We actually try very hard to maintain compatibility with Bash 3.2 for exactly the reason of MacOS X being stuck on Bash 3.2. Obviously for bugs without workarounds we have no choice, but we could add a note to the README for this known issue if we can prove its scope.

**Comment by augmentedfourth** [comment]
I updated my [liquidprompt 2 Homebrew PR](https://github.com/Homebrew/homebrew-core/pull/70536). Hopefully they like it an accept it this time.

**Comment by augmentedfourth** [comment]
2.0.2 Homebrew PR is accepted!

**Comment by Rycieos** [comment]
> > Would you like me to try upgrading bash?
> 
> Sure, we could isolate the issue to MacOS Bash 3.2, instead of MacOS Bash.

I once again have a MacOS laptop, so I can debug these types of things myself. Here is what I found:
* `/bin/bash` (version 3.2.57(1)-release (x86_64-apple-darwin20)) has this issue
* a freshly compiled bash 3.2 (version 3.2.57(1)-release (i386-apple-darwin20.6.0)) has this issue
* homebrew's precompiled binary (version 5.1.8(1)-release (x86_64--apple-darwin20.6.0)) does not have this issue
* a freshly compiled bash 4.0.28 (any lower patch has readline issues that prevent the prompt from being displayed) (version 4.0.28(2)-release (i386-apple-darwin20.6.0)) does not have this issue.

So whatever the bug is, it must be in 3.2, fixed in 4.0. Interestingly, it doesn't happen for me on Bash 3.2 on CentOS/Rocky with Cygwin. I scoured the CHANGELOG, and maybe these lines are related?
https://github.com/bminor/bash/blob/ce23728687ce9e584333367075c9deef413553fa/CHANGES#L4516
https://github.com/bminor/bash/blob/ce23728687ce9e584333367075c9deef413553fa/CHANGES#L4537
https://github.com/bminor/bash/blob/ce23728687ce9e584333367075c9deef413553fa/CHANGES#L4549
https://github.com/bminor/bash/blob/ce23728687ce9e584333367075c9deef413553fa/CHANGES#L4555
https://github.com/bminor/bash/blob/ce23728687ce9e584333367075c9deef413553fa/CHANGES#L4564

A number of those change descriptions mention "on some systems", so it seems likely that MacOS could have had a specific bug here. Bash changelogs never link specific commits (and Bash commit history is basically nonexistent anyway), so there is no way to know for sure.

So this seems to be 100% on Bash 3.2 here. This is not a breaking bug, just an annoyance, so we will continue (for now) to support Bash 3.2. If anyone discovers a workaround that we can apply in Liquidprompt, please let me know. But lacking such a fix, I'm closing this issue.

If anyone is having this issue, you have a large number of workarounds:
* Upgrade Bash (for example, using Homebrew) or use Zsh
* Configure the displayed character of the feature to be an ASCII character (for example, [`LP_MARK_BATTERY`](https://liquidprompt.readthedocs.io/en/stable/config.html#LP_MARK_BATTERY))
* Disable the feature that is printing a multi-byte character (like the Battery feature reported here)
* Ignore it. It's annoying, but not breaking.
